### DISCUSSION TITLE
Setting the include path for JIT compiler

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [CEED/libCEED#1695] [Q&A] [answered]
Hello,

I realized that during the JIT compilation of QFunctions, it always looks for the directory where the source is located to find headers. This incurs huge problems in complicated codes. Is there a way to set the include path for the JIT compilation?

**Comment by jeremylt** [comment, marked as answer]
You probably want 
https://libceed.org/en/latest/api/Ceed/#c.CeedAddJitSourceRoot

I'm not sure I understand what huge problems you forsee. Can you clarify?

**Reply by NamjaeChoi** [reply]
Thank you for the link. The problem occurs when the user derives their own object from the base framework object. The QFunction is defined in the same source with the user object definition, and thus it is exposed to all the headers in the framework which can be coming from multiple preset include directories.

Basically, we want to hide the QFunction from the user by providing them some macros that defines the necessary QFunctions automatically along with the shell of the user object, like:

```
#define CEED_KERNEL(name, parent)                                                                  \
  class name final : public parent<name>                                                           \
  {                                                                                                \
  public:                                                                                          \
    static CEEDQFunctionPointer _qf;                                                               \
    auto getQFunction() { return _qf; }                                                            \
    std::string getQFunctionPath() { return std::string(__FILE__) + ":" + #name + "_qf"; }
#define CEED_KERNEL_END(name)                                                                      \
  }                                                                                                \
  ;                                                                                                \
                                                                                                   \
  CEED_QFUNCTION(name##_qf)                                                                        \
  (void * ctx, const CeedInt Q, const CeedScalar * const * in, CeedScalar * const * out)           \
  {                                                                                                \
    auto kernel = static_cast<name *>(ctx);                                                        \
    kernel->computeResidualInternal(Q, in, out);                                                   \
    return 0;                                                                                      \
  }                                                                                                \
                                                                                                   \
  CEEDQFunctionPointer name::_qf = name##_qf;
 ```
 
Our QFunction is simply serving as the entry point, and the actual computation will be defined in the user device function (probably Kokkos function with the same backend with libCEED) that will be called by the entry QFunction.

**Reply by NamjaeChoi** [reply]
Thus the user class definition will look like:

```
CEED_KERNEL(CEEDDiffusionKernel, CEEDDiffusion)
public:
static InputParameters validParams();

CEEDDiffusionKernel(const InputParameters & parameters);
CEED_KERNEL_END(CEEDDiffusionKernel)
```

**Reply by jeremylt** [reply]
Hmm, we haven't tried anything like that yet, so I'm curious to see how it works out.

**Reply by NamjaeChoi** [reply]
Does `CeedAddJitSourceRoot` receive one directory at a time?

**Reply by jeremylt** [reply]
Correct. If you want to set multiple new root directories to search from, then you'd need to call this multiple times. Then you can specify the QFunction source path relative to any of the root dirs you've set.

**Reply by NamjaeChoi** [reply]
Thank you.

**Comment by NamjaeChoi** [comment]
Hmm does `CeedAddJitSourceRoot` only affect the relative path to the source? It doesn't seem to work with header include paths.

**Reply by jeremylt** [reply]
CeedAddJitSourceRoot sets a new root directory from which to search for QFunction source files.

So if `myqf.h` was in `foo/bar`, you could pass `foo/bar` as a new JiT source root and use `myqf.h` as the source path when creating the QFunction.

If a QFunction source file includes other source files, you'll want to use a relative path from the QFunction source file.

**Reply by NamjaeChoi** [reply]
`If a QFunction source file includes other source files`

This is the case we have and cannot use relative path, because the header being included also includes others that are from the framework. And those "others" are causing problems.

**Reply by jeremylt** [reply]
I'll be honest, this is really hard for me to follow without seeing the problem with my eyes.

Are you trying to use a source file that references other absolute paths? That *should* work ok I think? I can double check

**Reply by NamjaeChoi** [reply]
Let's say I have a QFunction defined at `CEEDDiffusion.h`. It includes `CEEDKernel.h` which is the framework header. `CEEDKernel.h` also includes a bunch of headers from the framework, and this nested inclusion continues. Setting all the nested inclusions as the relative path to the QFunction source file is not possible because those headers are also included elsewhere. Also they are not included using full path. The compiler resolves the include paths.

**Reply by jeremylt** [reply]
This sounds like C++ making our lives harder.

One way or another, we need the source path to the QFunction source. That then is read into a string buffer. Then, each include statement that is not a system header like math.h is replaced with the full contents of the file that is being included, recursively.

**Reply by NamjaeChoi** [reply]
So would it be possible to add an option to specify include paths of headers just like static compilers have the `-I` option to specify include paths, and do some search over the include directories, not just the source path?

**Reply by jeremylt** [reply]
Isn't that what setting multiple JiT root directories does?

**Reply by NamjaeChoi** [reply]
It doesn't set the include paths for the headers in the QFunction source, does it? It didn't work with my first trial. I understood that it only sets the search path for QFunction source itself.

As you said, `If a QFunction source file includes other source files, you'll want to use a relative path from the QFunction source file.`

**Reply by jeremylt** [reply]
Again, this is really hard for me to see what's happening without seeing the actual code.

Digging into the raw source on my side, I am guessing you are doing
```
#include <CEEDKernel.h>
```

If so, our JiT tools are reading that as a system header like math.h, stdio.h, etc and discarding it so the CUDA compiler doesn't get confused.

If that's what happening, I can make a PR to check all the root dirs provided to see if that include with <> in a file in those dirs? I'm uncertain if that will introduce some unexpected side effects but should be workable.

**Reply by NamjaeChoi** [reply]
No, it's being included like
```
#include "CEEDKernel.h"
```
and should not be ignored, and I can see it's indeed not ignored, as my JIT compilation is complaining about some headers that are deep in the framework.

**Reply by jeremylt** [reply]
Ok, now I think I understand with that little bit of code. That should be an easy fix.

**Reply by NamjaeChoi** [reply]
My include structure is: `CEEDDiffusion.h` -> `CEEDKernel.h` -> `CEEDKernelBase.h` -> `CEEDResidualObject.h` -> `GPUHeader.h` -> `Kokkos_Core.hpp`.

`Kokkos_Core.hpp` is in a different directory from the QFunction source (`CEEDDiffusion.h`) and cannot be located by the JIT compiler, which I would like to provide explicit include path.

**Reply by jeremylt** [reply]
> #include \"CEEDKernel.h\

Honestly, this was the most important part for me to trace where things could be going wrong.

**Reply by jeremylt** [reply]
Ok, branch is up. String parsing in C is tedious, but I should have a fix tomorrow? Maybe? This week definitely.

**Reply by NamjaeChoi** [reply]
Thank you for a prompt fix! I guess you will make `CeedAddJitSourceRoot` apply to the header include paths as well?

**Reply by jeremylt** [reply]
Right. I'll default to assuming a relative path in the current dir, and if the file is not found, I'll look in other JiT source dirs that have been set.

**Reply by jedbrown** [reply]
Just a nit here that `"header.h"` searches the current directory first, then behaves like `<header.h>`. (Technically, it's implementation-defined search, falling back to `<header.h>` semantics; with gcc there is also `-iquote`.) So if searching the current directory is not intended, I'd argue that include form should not be used.

Jeremy, more generally, I'm skeptical of the benefit of doing the string work ourselves versus just letting the compiler pick it up based on `-I` flags. Error messages are more clear when the compiler can report the original file:line.

**Reply by jeremylt** [reply]
Long term I'd love to move away from this string buffer business, but it's baked pretty deeply into how the JiT is setup currently and I couldn't guess how long it would take me to replace it.

**Reply by jeremylt** [reply]
See https://github.com/CEED/libCEED/pull/1696 for the patch that's WIP

**Reply by NamjaeChoi** [reply]
I guess people just use `"header.h"` and `<header.h>` interchangeably without deep thoughts, including me. I would love to see libCEED supporting both (i.e. `"header.h"` behaving like `<header.h>`). It's not feasible to change every occurrence of `"header.h"` to `<header.h>` in MOOSE. @lindsayad

**Reply by jeremylt** [reply]
*Some* people use them interchangeably, sure. We could support using `"header.h"` in the future for non-local headers, but my priority is making sure I don't break current functionality

**Reply by jeremylt** [reply]
A note here - a large number of transitive includes can really slow down the JiT, so I would include the absolute minimum number of files possible. Also, more includes is more chances to pull in something that CUDA/ROCm doesn't want to compile.

**Reply by lindsayad** [reply]
>  It's not feasible to change every occurrence of "header.h" to <header.h> in MOOSE.

It's not really feasible nor do I think it's very desirable. I think it's fairly common practice to use `"header.h"` when including files within your own library as opposed to a system or third-party library

**Reply by jeremylt** [reply]
Ideally, you should not be pulling in a very large number of transitive includes. The more stuff is in all those included files, the higher the likelihood that NVRTCC will refuse to compile.

**Reply by jedbrown** [reply]
I think libCEED should not try to force a special semantic, but in general, the semantic distinction is with whether the local directory is searched first, not a convention about whether the headers are packaged in "the same library" as the source that includes it. I'd say it's better (and presumably faster in the sense of fewer filesystem operations) to use `<header.h>` any time you don't deliberately want to search the local directory.

**Reply by jeremylt** [reply]
I don't know for certain, but I suspect my most recent commit to my branch will support either now? *But* I still recommend the QFunction source pulls in the absolute smallest number of transitive includes and lines of code that are strictly needed to compile the QFunction. The more that's pulled in, the more risk of unrelated changes breaking a QFunction. (This is why `ceed.h` was split and `ceed/types.h` was created: https://github.com/CEED/libCEED/blob/main/include/ceed/types.h `ceed/types.h` is recommended over `ceed.h.` in QFunction source)

**Reply by NamjaeChoi** [reply]
```I suspect my most recent commit to my branch will support either now?```
Yes, your recent change supports both.

**Comment by NamjaeChoi** [comment]
Hello,

I guess it would be great to have a feature to provide preprocessor definitions as well. The error we are seeing
```
/data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:125 in CeedCompile_Cuda(): NVRTC_ERROR_COMPILATION
/usr/include/gnu/stubs.h(7): catastrophic error: cannot open source file "gnu/stubs-32.h"
  # include <gnu/stubs-32.h>
```
is caused by the lack of compiler-provided preprocessor on the system architecture. `stubs.h` looks like:
```
#if !defined __x86_64__
# include <gnu/stubs-32.h>
#endif
#if defined __x86_64__ && defined __LP64__
# include <gnu/stubs-64.h>
#endif
#if defined __x86_64__ && defined __ILP32__
# include <gnu/stubs-x32.h>
#endif
```
I know your response would be not including this header, but that is not easily doable in the current MOOSE.

**Reply by jeremylt** [reply]
I have to be missing something. Why does the source file that should only include the physics code need to pull in almost every header related to MOOSE?

**Reply by NamjaeChoi** [reply]
Because the design pattern of MOOSE is to derive user class from base class and implement so-called "hook method" on the user's end, which is the physics code. It should see the full base class definition which pulls many framework headers.

**Reply by jeremylt** [reply]
There has to be ways to firewall that though and limit the total number of headers needed in JiT. For Ratel and HONEE, for example, we have headers that contain the bare minimal definitions and includes needed for JiT, and those minimal headers are included in the physics source files and the other headers that are used in the full compilation environment.

My concern is that by pulling every header on the system related to MOOSE into the JiT you'll end up feeding me a never ending stream of issues, and I've got a lot of high priority tasks on my plate right now. I can only work so fast, so I'll bottleneck your work if that's the case.

**Reply by jrwrigh** [reply]
[Misc googling ](https://stackoverflow.com/a/66047876/7564988)shows that NVRTC should provide the `__x86_64__` definitions. `CEED_DEBUG=1` might help find the issue; it'll output the raw text passed to NVRTC, but that obviously won't factor in the preprocessing done by NVRTC afterwards.

**Reply by NamjaeChoi** [reply]
Doesn't have to be fast. I'm just suggesting and I respect your priorities. I think I can somehow workaround for the time being by using function pointers even though hurting performance. @jrwrigh Thanks for the link.

**Reply by jeremylt** [reply]
https://github.com/CEED/libCEED/pull/1699

**Reply by NamjaeChoi** [reply]
Thank you for picking it up quickly!
